File: Programmers/level3/디스크 컨트롤러.cc
#include <bits/stdc++.h>
#define pp pair<int,int>

using namespace std;

priority_queue<pp, vector<pp>, greater<pp>> input_pq, job_pq;
int answer;
int cur_time;

void Select_Job() {
    auto job = job_pq.top();    job_pq.pop();

	if (job.second > cur_time) cur_time = job.second;

	int fin_time = cur_time + job.first;
	answer += (fin_time - job.second);
	cur_time = fin_time;
}

void Pq_Check() {
    if (job_pq.empty()) {
        auto job = input_pq.top(); input_pq.pop();
        job_pq.push({ job.second, job.first });
    }

    while (!input_pq.empty() and input_pq.top().first <= cur_time) {
        auto job = input_pq.top(); input_pq.pop();
        job_pq.push({ job.second, job.first });
    }
}

int solution(vector<vector<int>> jobs) {
    for (auto vec : jobs) {
        input_pq.push({ vec[0], vec[1] });
    }

    while (!input_pq.empty() or !job_pq.empty()) {
        Pq_Check();
        Select_Job();
    }

    answer /= jobs.size();
    return answer;
}

int main() {

    vector<vector<int>> jobs{ {0, 3} ,{1, 9},{2, 6} };
    cout << solution(jobs);
    return 0;
}

File: Programmers/level3/디스크 컨트롤러.java
import java.util.PriorityQueue;

class Solution {
    int readyTime;
    int curTime;
    PriorityQueue<Job> jobPq = new PriorityQueue<>();
    PriorityQueue<Ready> readyPq = new PriorityQueue<>();

    public int solution(int[][] jobs) {
        int jobCount = jobs.length;
        for (int[] job : jobs) {
            jobPq.add(new Job(job[0], job[1]));
        }

        while (!(jobPq.isEmpty() && readyPq.isEmpty())) {
            work();
            jobQMoveToReadyQ();
        }

        return readyTime / jobCount;
    }

    private void work() {
        if (readyPq.isEmpty()) {
            Job selectJob = jobPq.remove();
            curTime = selectJob.callTime + selectJob.needTime;
            readyTime += curTime - selectJob.callTime;
        } else {
            Ready selectJob = readyPq.remove();
            curTime += selectJob.needTime;
            readyTime += curTime - selectJob.callTime;
        }
    }

    private void jobQMoveToReadyQ() {
        while (!jobPq.isEmpty() && jobPq.element().callTime <= curTime) {
            readyPq.add(new Ready(jobPq.element().needTime, jobPq.element().callTime));
            jobPq.remove();
        }
    }
}

class Job implements Comparable<Job> {
    int callTime;
    int needTime;

    public Job(final int callTime, final int needTime) {
        this.callTime = callTime;
        this.needTime = needTime;
    }

    @Override
    public int compareTo(final Job o) {
        if (this.callTime == o.callTime) return Integer.compare(this.needTime, o.needTime);
        return Integer.compare(this.callTime, o.callTime);
    }
}

class Ready implements Comparable<Ready> {
    int needTime;
    int callTime;

    public Ready(final int needTime, final int callTime) {
        this.needTime = needTime;
        this.callTime = callTime;
    }

    @Override
    public int compareTo(final Ready o) {
        if (this.needTime == o.needTime) return Integer.compare(this.callTime, o.callTime);
        return Integer.compare(this.needTime, o.needTime);
    }
}
